SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp

```diff
@@ -0,0 +1,289 @@
+	for (FFileChangeData FileChange : FileChanges)
+	{
+		if (FileChange.Action == FFileChangeData::FCA_Added)
+		{
+			// Now we can start reading the new log file in the new log folder.
+			StartPollingLogFile(FPaths::Combine(FileChange.Filename, LaunchLogFilename));
+		}
+	}
+}
+
+void SSpatialOutputLog::ShutdownLogDirectoryWatcher(FString LogDirectory)
+{
+	AsyncTask(ENamedThreads::GameThread, [this, LogDirectory]
+	{
+		FDirectoryWatcherModule& DirectoryWatcherModule = FModuleManager::LoadModuleChecked<FDirectoryWatcherModule>(TEXT("DirectoryWatcher"));
+		if (IDirectoryWatcher * DirectoryWatcher = DirectoryWatcherModule.Get())
+		{
+			DirectoryWatcher->UnregisterDirectoryChangedCallback_Handle(LogDirectory, LogDirectoryChangedDelegateHandle);
+		}
+	});
+}
+
+void SSpatialOutputLog::CloseLogReader()
+{
+	if (GEditor != nullptr)
+	{
+		// Delete the old timer if one exists.
+		GEditor->GetTimerManager()->ClearTimer(PollTimer);
+	}
+
+	// Clean up the the previous file reader if it existed.
+	if (LogReader.IsValid())
+	{
+		LogReader->Close();
+		LogReader = nullptr;
+	}
+}
+
+void SSpatialOutputLog::StartPollingLogFile(FString LogFilePath)
+{
+	CloseLogReader();
+
+	// FILEREAD_AllowWrite is required as we must match the permissions of the other processes writing to our log file in order to read from it.
+	LogReader = CreateLogFileReader(*LogFilePath, FILEREAD_AllowWrite, PLATFORM_FILE_READER_BUFFER_SIZE);
+
+	if (LogReader.IsValid())
+	{
+		PollLogFile(LogFilePath);
+	}
+	else
+	{
+		UE_LOG(LogSpatialOutputLog, Error, TEXT("Could not set up log file reader for %s"), *LogFilePath);
+	}
+}
+
+void SSpatialOutputLog::PollLogFile(FString LogFilePath)
+{
+	// Poll log files in a background thread since we are doing a lot of string operations.
+	AsyncTask(ENamedThreads::AnyBackgroundThreadNormalTask, [this, LogFilePath]
+	{
+		if (!LogReader.IsValid())
+		{
+			UE_LOG(LogSpatialOutputLog, Error, TEXT("Attempted to read from log file but LogReader is not valid."));
+			return;
+		}
+
+		FScopedLoadingState ScopedLoadingState(*LogFilePath);
+
+		// Find out the current size of the log file. This is a cheaper operation than opening a new file reader on every poll.
+		LogReader->UpdateFileSize();
+
+		int32 SizeDifference = LogReader->TotalSize() - LogReader->Tell();
+
+		// New log lines have been added, serialize them.
+		if (SizeDifference > 0)
+		{
+			uint8* Ch = (uint8*)FMemory::Malloc(SizeDifference);
+
+			LogReader->Serialize(Ch, SizeDifference);
+
+			FString ReadResult;
+			FFileHelper::BufferToString(ReadResult, Ch, SizeDifference);
+
+			TArray<FString> LogLines;
+
+			// All log lines begin with 'time='. We use this as our log line delimiter.
+			ReadResult.ParseIntoArray(LogLines, TEXT("time="), true);
+
+			for (FString LogLine : LogLines)
+			{
+				FormatRawLogLine(LogLine);
+			}
+
+			FMemory::Free(Ch);
+		}
+
+		StartPollTimer(LogFilePath);
```

Comment:
How does this loop stop? The desctructor on this module doesn't seem to break it.

Reply:
It looks like we are trying to stop it by deleting the timer. Is that valid? What happens if you call set timer on a deleted timer?